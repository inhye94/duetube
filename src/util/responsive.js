export function checkMobile() {
  return window.innerWidth < 768 ? true : false;
}
