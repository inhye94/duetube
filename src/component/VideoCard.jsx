import React from "react";
import { FaCommentDots } from "react-icons/fa";
import { IoHeart, IoRocketSharp } from "react-icons/io5";
import { Link } from "react-router-dom";
import { formatAgo } from "../util/date";

export default function VideoCard({ video }) {
  const { title, channelTitle, publishedAt, thumbnails } = video.snippet;

  return (
    <article className="group w-full h-full px-[8px] py-[16px] rounded-xl">
      <Link
        className="flex flex-col-reverse h-full"
        title={title}
        to={`/watch/${video.id}`}
      >
        <div className="flex flex-col basis-full mt-[12px]">
          <h3 className="mb-[4px] text-[16px] font-bold line-clamp-2">
            {title}
          </h3>
          <strong className="mb-[12px] text-[14px] font-normal">
            {channelTitle}
          </strong>
          <p className="mt-auto text-[12px]">{formatAgo(publishedAt, "ko")}</p>

          {video.statistics && (
            <dl className="flex items-center gap-x-[6px] mt-[12px] text-[12px] opacity-70">
              <dt>
                <IoHeart aria-label="좋아요" />
              </dt>
              <dd>{setStatsFormat(video.statistics.likeCount)}</dd>
              <dt>
                <IoRocketSharp aria-label="조회수" />
              </dt>
              <dd>{setStatsFormat(video.statistics.viewCount)}</dd>
              <dt>
                <FaCommentDots aria-label="댓글수" />
              </dt>
              <dd>{setStatsFormat(video.statistics.commentCount)}</dd>
            </dl>
          )}
        </div>

        <div className="shrink-0 rounded-xl overflow-hidden group-hover:translate-x-[8px] group-hover:-translate-y-[8px] group-hover:border-solid group-hover:border-l-8 group-hover:border-b-8  border-purple-600 transition-all duration-300">
          <img
            className="aspect-video object-cover"
            src={thumbnails.medium.url}
            alt={`${title} 영상 썸네일`}
            loading="lazy"
          />
        </div>
      </Link>
    </article>
  );
}

const setStatsFormat = (value) => {
  if (!value) return 0;

  if (value.length < 3) {
    return value;
  } else if (value.length < 7) {
    return `${(value / 1000).toFixed(1)}천`;
  } else if (value.length < 11) {
    return `${(value / 1000000).toFixed(1)}만`;
  } else if (value.length < 15) {
    return `${(value / 1000000).toFixed(1)}억`;
  }
};

export { setStatsFormat };
