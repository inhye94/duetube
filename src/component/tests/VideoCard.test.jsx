import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import { Route, useLocation } from "react-router-dom";
import { withRouter } from "../../tests/utils";
import { mockVideo } from "../../tests/videos";
import { formatAgo } from "../../util/date";
import VideoCard from "../VideoCard";

describe("VideoCard", () => {
  const { title, channelTitle, publishedAt, thumbnails } = mockVideo.snippet;

  it("renders video items", () => {
    render(
      withRouter(<Route path="/" element={<VideoCard video={mockVideo} />} />)
    );

    const image = screen.getByRole("img");
    expect(image.src).toBe(thumbnails.medium.url);
    expect(image.alt).toBe(title + " 영상 썸네일");
    expect(screen.getByText(title)).toBeInTheDocument();
    expect(screen.getByText(channelTitle)).toBeInTheDocument();
    expect(screen.getByText(formatAgo(publishedAt, "ko"))).toBeInTheDocument();
  });

  it("navigates to detailed video page with video state when clicked", () => {
    // 테스트용 컴포넌트
    function LocationStateDisplay() {
      const location = useLocation();
      // VideoCard의 단위 테스트라 검증하는 컴포넌트를 mocking
      // 기능: VideoCard를 클릭하여 해당 '경로에 접근'하는지 + '정보를 전달'하는지 검증(이 컴포넌트는 정보 전달을 하지 않아서 pathname만 확인, state 전달은 useLocation으로 확인)
      return <pre>{location.pathname}</pre>;
    }

    render(
      withRouter(
        <>
          <Route path="/" element={<VideoCard video={mockVideo} />} />
          <Route
            path={`/watch/${mockVideo.id}`}
            element={<LocationStateDisplay />}
          />
        </>
      )
    );

    const card = screen.getByRole("link");
    userEvent.click(card);
    expect(screen.getByText(`/watch/${mockVideo.id}`)).toBeInTheDocument();
  });

  it("renders grid type correctly", () => {
    const { asFragment } = render(
      withRouter(<Route path="/" element={<VideoCard video={mockVideo} />} />)
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
