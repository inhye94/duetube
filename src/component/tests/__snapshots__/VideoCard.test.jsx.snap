// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoCard renders grid type correctly 1`] = `
<DocumentFragment>
  <article
    class="group w-full h-full px-[8px] py-[16px] rounded-xl"
  >
    <a
      class="flex flex-col-reverse h-full"
      href="/watch/12345"
      title="Test Video"
    >
      <div
        class="flex flex-col basis-full mt-[12px]"
      >
        <h3
          class="mb-[4px] text-[16px] font-bold line-clamp-2"
        >
          Test Video
        </h3>
        <strong
          class="mb-[12px] text-[14px] font-normal"
        >
          Test Channel
        </strong>
        <p
          class="mt-auto text-[12px]"
        >
          1년 전
        </p>
        <dl
          class="flex items-center gap-x-[6px] mt-[12px] text-[12px] opacity-70"
        >
          <dt>
            <svg
              aria-label="좋아요"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 448a32 32 0 01-18-5.57c-78.59-53.35-112.62-89.93-131.39-112.8-40-48.75-59.15-98.8-58.61-153C48.63 114.52 98.46 64 159.08 64c44.08 0 74.61 24.83 92.39 45.51a6 6 0 009.06 0C278.31 88.81 308.84 64 352.92 64c60.62 0 110.45 50.52 111.08 112.64.54 54.21-18.63 104.26-58.61 153-18.77 22.87-52.8 59.45-131.39 112.8a32 32 0 01-18 5.56z"
              />
            </svg>
          </dt>
          <dd>
            1.0천
          </dd>
          <dt>
            <svg
              aria-label="조회수"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M477.64 38.26a4.75 4.75 0 00-3.55-3.66c-58.57-14.32-193.9 36.71-267.22 110a317 317 0 00-35.63 42.1c-22.61-2-45.22-.33-64.49 8.07C52.38 218.7 36.55 281.14 32.14 308a9.64 9.64 0 0010.55 11.2l87.31-9.63a194.1 194.1 0 001.19 19.7 19.53 19.53 0 005.7 12L170.7 375a19.59 19.59 0 0012 5.7 193.53 193.53 0 0019.59 1.19l-9.58 87.2a9.65 9.65 0 0011.2 10.55c26.81-4.3 89.36-20.13 113.15-74.5 8.4-19.27 10.12-41.77 8.18-64.27a317.66 317.66 0 0042.21-35.64C441 232.05 491.74 99.74 477.64 38.26zM294.07 217.93a48 48 0 1167.86 0 47.95 47.95 0 01-67.86 0z"
              />
              <path
                d="M168.4 399.43c-5.48 5.49-14.27 7.63-24.85 9.46-23.77 4.05-44.76-16.49-40.49-40.52 1.63-9.11 6.45-21.88 9.45-24.88a4.37 4.37 0 00-3.65-7.45 60 60 0 00-35.13 17.12C50.22 376.69 48 464 48 464s87.36-2.22 110.87-25.75A59.69 59.69 0 00176 403.09c.37-4.18-4.72-6.67-7.6-3.66z"
              />
            </svg>
          </dt>
          <dd>
            5.0천
          </dd>
          <dt>
            <svg
              aria-label="댓글수"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 32C114.6 32 0 125.1 0 240c0 49.6 21.4 95 57 130.7C44.5 421.1 2.7 466 2.2 466.5c-2.2 2.3-2.8 5.7-1.5 8.7S4.8 480 8 480c66.3 0 116-31.8 140.6-51.4 32.7 12.3 69 19.4 107.4 19.4 141.4 0 256-93.1 256-208S397.4 32 256 32zM128 272c-17.7 0-32-14.3-32-32s14.3-32 32-32 32 14.3 32 32-14.3 32-32 32zm128 0c-17.7 0-32-14.3-32-32s14.3-32 32-32 32 14.3 32 32-14.3 32-32 32zm128 0c-17.7 0-32-14.3-32-32s14.3-32 32-32 32 14.3 32 32-14.3 32-32 32z"
              />
            </svg>
          </dt>
          <dd>
            0.2천
          </dd>
        </dl>
      </div>
      <div
        class="shrink-0 rounded-xl overflow-hidden group-hover:translate-x-[8px] group-hover:-translate-y-[8px] group-hover:border-solid group-hover:border-l-8 group-hover:border-b-8  border-purple-600 transition-all duration-300"
      >
        <img
          alt="Test Video 영상 썸네일"
          class="aspect-video object-cover"
          loading="lazy"
          src="https://example.com/thumbnail.jpg"
        />
      </div>
    </a>
  </article>
</DocumentFragment>
`;
